### ISSUE TITLE
[clang-tidy] False positive with cppcoreguidelines-narrowing-conversions from a comparison operator.

### ISSUE THREAD

**Issue by brackeen** [llvm/llvm-project#62123]
On macOS, the `BOOL` type is a `signed char`, which results in some narrowing conversion warnings.

test.c:
```c
typedef signed char BOOL;

BOOL foo(int i) {
    BOOL result = (i == 0);
    return result;
}
```

Running (clang-tidy 16.0.1):

```
clang-tidy -checks=cppcoreguidelines-narrowing-conversions test.c
```

Result:

```console
test.c:4:19: warning: narrowing conversion from 'int' to signed type 'BOOL' (aka 'signed char') is implementation-defined [cppcoreguidelines-narrowing-conversions]
    BOOL result = (i == 0);
                  ^
```

In C++, a similar warning:
```console
test.cc:4:19: warning: narrowing conversion from 'bool' to signed type 'BOOL' (aka 'signed char') is implementation-defined [cppcoreguidelines-narrowing-conversions]
    BOOL result = (i == 0);
                  ^
```

**Comment by mclow** [comment]
In C++, it is in fact a narrowing conversion - a `bool` is 4 bytes long, and a `BOOL` is only one byte long.
It would be nice, though if clang was aware enough that the expression `(i == 0)` is only going to produce values of 0 or 1, and not warn.
